server/rest/src/main/java/org/infinispan/rest/resources/CounterResource.java

```diff
@@ -30,84 +50,299 @@
+
+         return response.thenApply(res -> responseBuilder.entity(Long.toString(res)).build());
+      });
    }
 
-   private MediaType negotiateMediaType(String accept) {
-      return MediaType.parseList(accept).filter(t ->
-            t.match(MediaType.TEXT_PLAIN) // TODO: add more types in ISPN-10211
-      ).findFirst().orElseThrow(() -> Log.REST.unsupportedDataFormat(accept));
+   private CompletionStage<Long> getWeakCounterValue(String weakCounterName) {
+      return CompletableFuture.supplyAsync(() -> counterManager.getWeakCounter(weakCounterName).getValue(),
+            invocationHelper.getExecutor());
    }
 
+   private CompletionStage<RestResponse> resetCounter(RestRequest request) throws RestResponseException {
+      String counterName = request.variables().get("counterName");
+
+      return counterManager.getConfigurationAsync(counterName).thenCompose(configuration -> {
+         if (configuration == null) return completedFuture(new NettyRestResponse.Builder().status(NOT_FOUND).build());
+         CompletionStage<Void> result = configuration.type() == CounterType.WEAK ?
+               counterManager.getWeakCounter(counterName).reset() :
+               counterManager.getStrongCounter(counterName).reset();
 
-   private CompletionStage<RestResponse> addCounter(RestRequest request) {
+         return result.thenApply(v -> new NettyRestResponse.Builder().build());
+      });
+   }
+
+   private CompletionStage<RestResponse> incrementWeak(RestRequest request) {
+      return executeWeakCounterOp(request, WeakCounter::increment);
+   }
+
+   private CompletionStage<RestResponse> decrementWeak(RestRequest request) {
+      return executeWeakCounterOp(request, WeakCounter::decrement);
+   }
+
+   private CompletionStage<RestResponse> addValueWeak(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
       String counterName = request.variables().get("counterName");
-      NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+      Long delta = checkForNumericParam("delta", request, responseBuilder);
+      if (delta == null) return completedFuture(responseBuilder.build());
 
-      CounterConfiguration configuration = counterManager.getConfiguration(counterName);
-      if (configuration == null) {
-         return CompletableFuture.completedFuture(builder.status(HttpResponseStatus.NOT_FOUND.code()).entity("Counter not found").build());
-      }
-      builder.header(HttpHeaderNames.CACHE_CONTROL.toString(), CacheControl.noCache());
+      CompletionStage<WeakCounter> weakCounter = getWeakCounter(counterName);
+      return weakCounter.thenCompose(counter -> {
+         if (counter == null) {
+            responseBuilder.status(HttpResponseStatus.BAD_REQUEST).entity(String.format("Weak counter '%s' not found", counterName));
+            return completedFuture(responseBuilder.build());
+         }
+         counter.add(delta);
```

Comment:
shouldn't wait for the `add()` to complete before sending the reply?

Reply:
Yes!